**Финальный проект**

**Задания по SQL**
_____
**Описание исследования.**
    Исследование приложения для чтения книг по подписке. 
_____
**Цель исследования.**
    Провести исследование данных о книжном сервисе с целью сформулировать ценностное предложение для нового продукта.
   
_____
**Задачи исследования.**

    Необходимо провести исследования базы данных, вывести и проанализировать информацию о книгах, издательствах,авторах и пользовательских обзоров.
_____
**Исходные данные.**

Таблица books
    Содержит данные о книгах:
    book_id — идентификатор книги;
    autor_id — идентификатор автора;
    title - название книги;
    num_pages - количество страниц;
    publication_date — дата публикации книги;
    publisher_id — идентификатор издателя.
    
    Таблица authors
    Содержит данные об авторах:
    author_id - идентификатор автора;
    author —	имя автора.

    Таблица publishers
    Содержит данные об издательствах:
    publisher_id —	идентификатор издательства;
    publisher —	название издательства;

    Таблица raitings
    Содержит данные о пользовательских оценках книг:
    raiting_id —	идентификатор оценки;
    book_id — идентификатор книги;
    username — имя пользователя, оставившего оценку;
    raiting — оценка книги.
 
    Таблица reviews
    Содержит данные о пользовательских обзорах на книги:
    review_id — идентификатор обзора;
    book_id — идентификатор книги;
    username — имя пользователя, написавшего обзор;
    text — текст обзора.


**Содержание работы**

**[Часть 1. Загрузка и изучение данных.](#1-bullet)**

**Часть 2. Задачи**
* [1. Посчитайте, сколько книг вышло после 1 января 2000 года.](#2-bullet)
* [2. Для каждой книги посчитайте количество обзоров и среднюю оценку.](#3-bullet)
* [3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц.](#4-bullet)
* [4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.](#5-bullet)
* [5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.](#6-bullet)

 **[Часть 3. Вывод](#7-bullet)**

In [1]:
!pip install psycopg2-binary

In [2]:
# импортируем библиотеки
import pandas as pd
import sqlalchemy as sa

<a id = '1-bullet'></a>
### Часть 1. Загрузка и изучение данных.

In [3]:
# устанавливаем параметры
db_config = {
            'user': 'praktikum_student', # имя пользователя
            'pwd': 'Sdf4$2;d-d30pp', # пароль
            'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
            'port': 6432, # порт подключения
            'db': 'data-analyst-final-project-db' # название базы данных
            }
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = sa.create_engine(connection_string, connect_args={'sslmode':'require'})

# чтобы выполнить SQL-запрос, пишем функцию с использованием Pandas
def get_sql_data(query:str, engine:sa.engine.base.Engine=engine) -> pd.DataFrame:
    with engine.connect() as con:
        return pd.read_sql(sql=sa.text(query), con = con)


In [4]:
# формируем запрос и выводим данные по таблице о книгах
query = '''SELECT * 
           FROM books 
           LIMIT 5'''

get_sql_data(query)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
# выводим количество записей в таблице о книгах
query = '''SELECT  COUNT(*)
           FROM books '''

get_sql_data(query)

count
0   1000

In [6]:
# выводим данные по таблице о издательствах
query = '''SELECT * 
           FROM publishers 
           LIMIT 5'''

get_sql_data(query)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [7]:
# выводим количество записей в таблице о издательствах
query = '''SELECT COUNT(*) 
           FROM publishers '''

get_sql_data(query)

count
0    340

In [8]:
# формируем запрос и выводим данные по таблице об авторах
query = '''SELECT * 
           FROM authors 
           LIMIT 5'''

get_sql_data(query)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [9]:
# выводим количество записей в таблице об авторах
query = '''SELECT  COUNT(*)
           FROM authors '''

get_sql_data(query)

count
0    636

In [10]:
# выводим данные по таблице о рейтингах
query = '''SELECT * 
           FROM ratings 
           LIMIT 5'''

get_sql_data(query)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [11]:
# выводим количество записей в таблице о рейтингах
query = '''SELECT  COUNT(*)
           FROM ratings '''

get_sql_data(query)

count
0   6456

In [12]:
# выводим данные по таблице об обзорах на книги
query = '''SELECT * 
           FROM reviews 
           LIMIT 5'''

get_sql_data(query)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [13]:
# выводим количество записей в таблице об обзорах на книги
query = '''SELECT  COUNT(*)
           FROM reviews '''

get_sql_data(query)

count
0   2793

В данных имеем 5 таблиц. Таблица с книгами содержит 1000 записей. Таблица с авторами содержит 636 записей. Таблица с издательствами содержит 340  записей. Количество записей в таблице с рейтингами по книгам 6456. Количество записей в таблице отзывов на книги 2793.

### Часть 2. Задачи

<a id = '2-bullet'></a>
#### Посчитайте, сколько книг вышло после 1 января 2000 года.

In [14]:
query = '''SELECT  COUNT(book_id)
           FROM books
           WHERE CAST(publication_date AS date) > '2000-01-01' '''

get_sql_data(query)

count
0    819

Общее количество книг вышедшее после 1 января 2000 (без учета 1 января 2000) составляет 819 книг.

<a id = '3-bullet'></a>
#### Для каждой книги посчитайте количество обзоров и среднюю оценку.

In [15]:
query = '''SELECT  b.title,
                   COUNT(DISTINCT(rv.review_id)) AS count_reviews,
                   ROUND(AVG(rt.rating), 2) AS average_rating
           FROM books AS b LEFT JOIN ratings AS rt ON b.book_id=rt.book_id
           LEFT JOIN reviews AS rv ON b.book_id=rv.book_id
           GROUP BY b.book_id
           ORDER BY count_reviews DESC
           LIMIT 5 '''

get_sql_data(query)

title  count_reviews  \
0                            Twilight (Twilight  #1)              7   
1                          Outlander (Outlander  #1)              6   
2                                    Eat  Pray  Love              6   
3  Harry Potter and the Chamber of Secrets (Harry...              6   
4  Harry Potter and the Prisoner of Azkaban (Harr...              6   

   average_rating  
0            3.66  
1            4.13  
2            3.40  
3            4.29  
4            4.41

Посчитали количество отзывов и средние оценки по книгам. Самое большое количество отзывов у книги Сумерки - 7 отзывов. 

<a id = '4-bullet'></a>
#### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц.

In [16]:
query = '''SELECT p.publisher,
                  COUNT(b.book_id)
           FROM publishers AS p JOIN books AS b ON p.publisher_id=b.publisher_id
           WHERE b.num_pages>50
           GROUP BY p.publisher_id
           ORDER BY COUNT(b.book_id) DESC
           LIMIT 1 '''

get_sql_data(query)

publisher  count
0  Penguin Books     42

Издательство выпустившее больше всего книг толще 50 страниц это Penguin Books, всего 42 книги.

<a id = '5-bullet'></a>
#### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.

In [18]:
query = '''SELECT a.author,
                  ROUND(AVG(r.rating), 4) AS book_rating
            FROM books b JOIN authors a ON b.author_id=a.author_id
            JOIN ratings r ON b.book_id=r.book_id
            WHERE b.book_id IN (SELECT b.book_id
                                FROM books AS b JOIN ratings AS rt ON b.book_id=rt.book_id
                                GROUP BY b.book_id HAVING COUNT(rt.rating_id)>50)
            GROUP BY a.author_id
            ORDER BY book_rating DESC
            LIMIT 1 '''

get_sql_data(query)

author  book_rating
0  J.K. Rowling/Mary GrandPré       4.2871

Автором с самой высокой средней оценкой книг является J.K. Rowling/Mary GrandPré.

<a id = '6-bullet'></a>
####  Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [19]:
query = '''WITH counts AS (SELECT COUNT(review_id) OVER (PARTITION BY username) AS count_review
                           FROM reviews
                           WHERE username IN (SELECT username
                                              FROM ratings 
                                              GROUP BY username
                                              HAVING COUNT(rating_id)>48))
           SELECT ROUND(AVG(count_review)) AS avg_counts_review
           FROM counts '''

get_sql_data(query)

avg_counts_review
0               24.0

Среднее количество обзоров от пользователей, поставивших более 48 оценок составило 24 штуки.

<a id = '7-bullet'></a>
### Выводы.

В ходе анализа имеющихся данных изучили 5 таблиц. Таблица с книгами содержит 1000 записей. Таблица с авторами содержит 636 записей. Таблица с издательствами содержит 340  записей. Количество записей в таблице с рейтингами по книгам 6456. Количество записей в таблице отзывов на книги 2793.
Общее количество книг вышедшее после 1 января 2000 (без учета 1 января 2000) составляет 819 книг.
Посчитали количество отзывов и средние оценки по книгам. Самое большое количество отзывов у книги Сумерки - 7 отзывов. 
Издательство выпустившее больше всего книг толще 50 страниц это Penguin Books, всего 42 книги.Автором с самой высокой средней оценкой книг является J.K. Rowling/Mary GrandPré.
Среднее количество обзоров от пользователей, поставивших более 48 оценок составило 24 штуки.